alf/experience_replayers/replay_buffer.py

```diff
@@ -588,40 +588,80 @@ def gather_all(self, ignore_earliest_frames: bool = False):
+        end_idx = self.circular(pos)
+
+        if start_idx < end_idx:
+            # This is the happy case when
+            #
+            # |---o**********************o------| env 0
+            # |---o**********************o------| env 1
+            # |             .                   |
+            # |             .                   |
+            # |             .                   |
+            # |---o**********************o------| env n-1
+            #     ^                      ^
+            #  start_idx              end_idx
+            #
+            # We just collect the "*"s in the natural order.
+            result = alf.nest.map_structure(
+                lambda buf: buf[:, start_idx:end_idx, ...], self._buffer)
         else:
-            if ignore_earliest_frames:
-                actual_start = self._num_earliest_frames_ignored
-            else:
-                actual_start = 0
-            # Assumes that non-full buffer always stores data starting from 0
+            # This is the more complicated case when
+            #
+            # |#########o---o*******************| env 0
+            # |#########o---o*******************| env 1
+            # |             .                   |
+            # |             .                   |
+            # |             .                   |
+            # |#########o---o*******************| env n-1
+            #           ^   ^
+            #     end_idx   start_idx
+            #
+            # In this case we need to collect the "*"s first and then "#"s.
+            #
+            # Note that the start_idx and end_idx might does not have to be
+            # adjacent to each other because it may ignore_earliest_frames.
             result = alf.nest.map_structure(
-                lambda buf: buf[:, actual_start:size, ...], self._buffer)
-            info = BatchInfo(
-                env_ids=torch.arange(self._num_envs),
-                positions=(torch.ones(self._num_envs, dtype=torch.int64) *
-                           actual_start))
-            if alf.get_default_device() != self._device:
-                result, info = convert_device((result, info))
-            info = info._replace(replay_buffer=self)
+                lambda buf: torch.cat(
+                    (buf[:, start_idx:, ...], buf[:, :end_idx, ...]), dim=1),
+                self._buffer)
+
+        info = BatchInfo(
+            env_ids=torch.arange(self._num_envs),
+            positions=(
+                torch.ones(self._num_envs, dtype=torch.int64) * start_pos))
+
+        # TODO(breakds): Maybe remove this in a future PR. It is not desired to
+        # move everything to the GPU memory yet. It will be more flexiable (and
+        # GPU memory friendly) to do this in a later stage where we only need to
+        # store a mini batch in the GPU memory at a atime.
+        if alf.get_default_device() != self._device:
+            result, info = convert_device((result, info))
```

Comment:
If the caller really want the data on CPU, it can always call this function under CPU context using `with alf.device("cpu")`

Reply:
Makes a lot of sense. Did not know this convention.
I removed the `TODO` comments.